runtime/lua/vim/_memoize.lua

```diff
@@ -0,0 +1,51 @@
+--- Module for private utility functions
```

Comment:
Although this is private for now, I'd encourage putting it in `vim.func` since that is where we would likely want to expose it eventually. This is on my mind since we're also considering `traverse()`, which would also make sense in `vim.func`.